rlp/rlp.go

```diff
@@ -0,0 +1,100 @@
+package rlp
+
+import (
+	"encoding/binary"
+	"math"
+)
+
+type Item struct {
+	D []byte
+	L []*Item
+}
+
+func Encode(input *Item) []byte {
+	if input.D != nil {
+		if len(input.D) == 1 && input.D[0] < 128 {
+			return input.D
+		}
+		return append(encodeLength(input.D, 128), input.D...)
+	}
+
+	var out []byte
+	for i := range input.L {
+		out = append(out, Encode(input.L[i])...)
+	}
+	return append(encodeLength(out, 192), out...)
+}
+
+func encodeLength(input []byte, offset uint8) []byte {
+	switch l := uint64(len(input)); {
+	case l < 56:
+		header := uint64(offset) + l
+		return []byte{byte(header)} // (max(128,192) + 55) < 255
+	case l <= math.MaxUint64:
+		var (
+			buf    = make([]byte, binary.MaxVarintLen64)
+			n      = binary.PutUvarint(buf, l) //bytes needed to encode length
+			header = int(offset) + 55 + n
+		)
+		return append([]byte{byte(header)}, buf[:n]...)
+	default:
+		return []byte{}
+	}
+}
+
+func Decode(input []byte) *Item {
+	switch {
+	case input[0] < 128: // string
+		return &Item{D: []byte{input[0]}}
+	case input[0] < 184: // string
+		return &Item{D: input[1:]}
```

Comment:
Just a quick suggestion here to cap the length of the slice by doing something like `datalen := input[0]` and `return &Item{D: input[1:datalen]}`

Reply:
oh interesting. are you thinking of a case where the actual data is longer than its specified length?

Reply:
Yeah or at least to explicitly guard against a case where the length prefix was malformed

Reply:
OK. I think we have this nailed down in the latest commit